blog-post/blog-post.md

```diff
@@ -0,0 +1,738 @@
+    let b1 : bool = toBool(and(shr(1, bools), 0x1));
+    let b2 : bool = toBool(and(shr(2, bools), 0x1));
+    return fn(b0, b1, b2);
+}
+```
+
+Function `unpack_bools` extracts three individual boolean values from the lowest three bits
+of a packed `word` (`bools`). It treats it as a bitmask where the least significant bit
+(position 0) represents `b0`, the next bit (position 1) represents `b1`, and the third bit
+(position 2) represents `b2`. After extracting these bits and converting them to booleans,
+it passes them as arguments to a callback function `fn` and returns whatever result
+that function produces.
+
+We also support the definition of (non-recursive) anonymous functions using the `lam` keyword.
+Functions defined in this way can capture values available in the defining scope. As an example
+consider this testing utility that counts the number of times an arbitrary function is called:
+
+```solidity
+forall function count_calls(fn : (T) -> U) -> (memory(word), (T) -> U) {
+    let counter : memory(word) = allocate(32);
+    return (counter, lam (a : T) -> {
+        counter += 1;
+        return fn(a);
+    };
+}
+```
+
+Our implementation here is similar to systems languages like Rust and C++: the compiler produces a
+unique type for each anonymous function that contains the captured state, and these unique types are
+made callable by making them instances of the built in `invokable` type class (similar to the `Fn`
+trait in Rust). This approach avoids the kind of dynamic heap based environment management used in
+higher level languages like Haskell or Python.
+
+### Type inference
+
+Core Solidity supports the inference of types in almost any position. Annotations are usually optional and
+need be added only where considered desirable for readability or understanding. Inference is decidable,
+and the situations in which ambiguities requiring annotation can occur are very limited. This lets
+us solve a lot of the syntactic clutter required when writing Solidity today. As an example,
+consider the following Classic Solidity definition:
+
+```solidity
+struct LendingPosition {
+    uint256 collateralDeposited;
+    uint256 debtIssued;
+    uint256 lastUpdated;
+}
+```
+
+Instantiating `LendingPosition` in Classic Solidity is syntactically clunky, requiring us
+to both annotate the type and invoke the `LendingPosition` constructor:
+
+```solidity
+LendingPosition emptyPosition = LendingPosition(0, 0, block.timestamp);
+```
+
+Since we are using the `LendingPosition` constructor, the compiler can infer
+the type of `emptyPosition`. Core Solidity allows us to omit the redundant annotation:
+
+```solidity
+let emptyPosition = LendingPosition(0, 0, block.timestamp);
+```
+
+Assigning an expression to a variable in Classic can often result in redundant annotation if those
+types are already present in the expression being assigned:
+
+```solidity
+(uint256 a, uint256 b) = abi.decode(input, (uint256, uint256))
```

Comment:
The repetitiveness would be even more apparent IMO if you used something that has a location, e.g.:

```suggestion
(bytes memory a, bytes memory b) = abi.decode(input, (bytes, bytes))
```